src/heap_roots.h

```diff
@@ -31,54 +31,71 @@ typedef LinkedFifo<FinalizerNode> FinalizerNodeFifo;
 
 class FinalizerNode : public FinalizerNodeFifo::Element {
  public:
+  FinalizerNode(HeapObject* key, ObjectHeap* heap) : key_(key), heap_(heap) {}
   virtual ~FinalizerNode();
-  virtual bool has_key(HeapObject* value) = 0;
 
+  // Called at the end of compaction and at other times where all pointers
+  // should be visited with no weakness/finalization processing.
   virtual void roots_do(RootCallback* cb) = 0;
-  // Should return false if the node needs GC processing.
-  virtual bool alive(LivenessOracle* oracle) = 0;
-  // Should return null if the node should be deleted.
-  virtual bool handle_not_alive(RootCallback* process_slots, ObjectHeap* heap) = 0;
+  // Cleanup when a heap is deleted.
+  virtual void heap_dying() {}
+  // Should return true if the node should be unlinked.
+  virtual bool weak_processing(bool in_closure_queue, RootCallback* visitor, LivenessOracle* oracle) = 0;
+
+ protected:
+  HeapObject* key_;
+  ObjectHeap* heap_;
 };
 
-class ToitFinalizerNode : public FinalizerNode {
+class CallableFinalizerNode : public FinalizerNode {
  public:
-  ToitFinalizerNode(HeapObject* key, Object* lambda)
-    : key_(key), lambda_(lambda) {}
+  CallableFinalizerNode(HeapObject* key, Object* lambda, ObjectHeap* heap)
+    : FinalizerNode(key, heap), lambda_(lambda) {}
 
-  HeapObject* key() { return key_; }
-  void set_key(HeapObject* value) { key_ = value; }
-  virtual bool has_key(HeapObject* value);
   Object* lambda() { return lambda_; }
```

Comment:
const

Reply:
No, we can modify the object via this call.